src/Functions/FunctionDateOrDateTimeAddInterval.h

```diff
@@ -45,140 +47,104 @@ struct AddNanosecondsImpl
 {
     static constexpr auto name = "addNanoseconds";
 
-    static inline NO_SANITIZE_UNDEFINED DecimalUtils::DecimalComponents<DateTime64>
-    execute(DecimalUtils::DecimalComponents<DateTime64> t, Int64 delta, const DateLUTImpl &, UInt16 scale = DataTypeDateTime64::default_scale)
-    {
-        Int64 multiplier = DecimalUtils::scaleMultiplier<DateTime64>(9 - scale);
-        auto division = std::div(t.fractional * multiplier + delta, static_cast<Int64>(1000000000));
-        return {t.whole * multiplier + division.quot, t.fractional * multiplier + delta};
-    }
-
     static inline NO_SANITIZE_UNDEFINED DateTime64
     execute(DateTime64 t, Int64 delta, const DateLUTImpl &, UInt16 scale = 0)
     {
         Int64 multiplier = DecimalUtils::scaleMultiplier<DateTime64>(9 - scale);
-        return t * multiplier + delta;
+        return DateTime64(DecimalUtils::multiplyAdd(t.value, multiplier, delta));
     }
 
-    static inline NO_SANITIZE_UNDEFINED UInt32 execute(UInt32 t, Int64 delta, const DateLUTImpl &, UInt16 = 0)
+    static inline NO_SANITIZE_UNDEFINED DateTime64 execute(UInt32 t, Int64 delta, const DateLUTImpl &, UInt16 = 0)
     {
         Int64 multiplier = DecimalUtils::scaleMultiplier<DateTime64>(9);
-        return static_cast<UInt32>(t * multiplier + delta);
+        return DateTime64(DecimalUtils::multiplyAdd(static_cast<Int64>(t), multiplier, delta));
     }
 
-    static inline NO_SANITIZE_UNDEFINED DateTime64 execute(UInt16, Int64, const DateLUTImpl &, UInt16 = 0)
+    static inline NO_SANITIZE_UNDEFINED Int8 execute(UInt16, Int64, const DateLUTImpl &, UInt16 = 0)
     {
-        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "addNanoSeconds() cannot be used with Date");
+        chassert(false);
+        return 0;
```

Comment:
maybe throw a logical error in cases like this?
Returning 0 seems strange on something that shouldn't be called + it could make it harder to debug.

Reply:
Yeah, that sounds better, changing.